### ISSUE TITLE
Lerna doesn't work with NPM automation tokens

### ISSUE THREAD

**Issue by bcomnes** [lerna/lerna#2788]
When I properly set up npm credentials inside of GitHub Actions, and use an npm automation token (which normally supports write actions), lerna fails when verifying npm auth, even though npm is properly authenticated for publishing:

```
$ lerna publish patch
....
lerna info Verifying npm credentials
lerna http fetch GET 403 https://registry.npmjs.org/-/npm/v1/user 241ms
403 Forbidden - GET https://registry.npmjs.org/-/npm/v1/user
lerna ERR! EWHOAMI Authentication error. Use `npm whoami` to troubleshoot.
```

When I switch to a read-write npm token (with OTP turned off), lerna works fine. 

It appears lerna does not support npm automation tokens.  It would be nice if it did, that way people won't be inclined to turn off OTP on their human tokens.

**Comment by JamesHenry** [comment]
I just hit this as well, I am imagine these automation tokens are very new, as I don't remember that being the experience on NPM before now, but yeah... lerna is not compatible with them right now

**Comment by bcomnes** [comment]
I was surprised there were no other issues (that I could locate) about this, since they have been out for a while.  My feeling is that lerna uses the token to do reads about the user and the automation token only supports writes?  Just a guess though.

**Comment by grug** [comment]
We just spent the day trying to work out why our publish command wasn't working and it turned out it was this! The NPM output wasn't a bit more descriptive to let you know that X permission was missing (although I guess that might be for security reasons)

**Comment by dyladan** [comment]
Using <https://github.com/lerna/lerna/tree/f6e7a13e60fefc523d701efddfcf0ed41a77749b/commands/publish#--no-verify-access> may fix this. Automation tokens don't have permission to access the endpoint lerna uses to verify permission.

As a longer-term fix, this should be documented or the `/-/whoami` endpoint should be used which can be accessed by automation tokens.

**Comment by bcomnes** [comment]
Oh good catch.  I can't remember if I tried that or not.  Anyone else try?

**Comment by ling1726** [comment]
This also happened recently for us, we scratched our heads for a while before we suspected it was the `npm/v1/user` endpoint that was failing.

adding `--no-verify-access` to bypass that check works, and if the token doesn't have publish permissions then you get the default 404 npm error.

**Comment by bcomnes** [comment]
Thank you for confirming @ling1726 I will be trying that.

I still think handling could be improved here, though its good to know that work around works.

**Comment by FRSgit** [comment]
This issue thread was a real life saver - I've just struggled with publishing the [@frsource/tiny-carousel](https://github.com/FRSOURCE/tiny-carousel) packages for two days before finding this thread.
To make life easier for the future travellers, I've patched a PR adding the necessary info about handling of npm automation tokens - #2825. Hope it will save some time for somebody 😄 
Feel free to comment/post suggestions in the PR, or react with 👍  if agree that it should be merged.

**Comment by GiancarlosIO** [comment]
Yep, it would be great if the docs can explain that Lerna doesn't support automation npm token 🙏

**Comment by gziolo** [comment]
I run into the same issue while automating npm publishing with GitHub action in https://github.com/WordPress/gutenberg/pull/39259. Thank you for the hint with `--no-verify-access` 👍🏻

**Comment by JamesHenry** [comment]
Now that Nrwl has taken over stewardship we are going to take a look into this issue!

As was already mentioned in the comments here, you can use `"verifyAccess": false` in your lerna.json publish config, or pass `--no-verify-access` to the `lerna publish` command to get it to skip the auth step, which unblocks this issue for now.

It is likely that the logic which validates the npm auth piece just needs to be expanded/fixed and as mentioned we will look into that very soon as we also want to address the highly related: https://github.com/lerna/lerna/issues/2730

**Comment by JamesHenry** [comment]
**TL;DR - automation tokens will just work as expected in 5.2.0 of lerna and later**

---

Hi folks, we just released `5.2.0` of lerna: https://github.com/lerna/lerna/releases/tag/v5.2.0

Amongst other things, in this version we set the legacy `verifyAccess` behavior to be false by default. It will still respect if you set it to be true explicitly.

Essentially, prior to `5.2.0` lerna would, by default, make some preemptive API calls to npm using the token you provide to try and "fast fail" for auth concerns. This approach made sense when it was implemented, but there are now multiple types of tokens supported by npm (and perhaps more in the future), and they have different access rights. That's what made the previous behavior break for automation tokens, and was the underlying cause of this issue.

Given that the npm API requests will fail with appropriate auth errors as part of the publishing process anyway, and that it is not something which scales to all token types, we took the decision to invert the `verifyAccess` logic and only enable when a user explicitly wants it.

Hope that all makes sense, this issue can now be marked as resolved!

**Comment by rbirkgit** [comment]
I am still getting 403 errors with lerna 5.2.0:

https://github.com/lerna/lerna/issues/3217

**Comment by erickayalarias** [comment]
someone know how to fix this issue? Im still having this issue but with Lerna v3 everything was working fine, i had tried almost every solution but with no success

**Comment by JamesHenry** [comment]
@erickayalarias As per the note above, lerna does not perform any access verification by default.

Ensure that you are using a modern version of lerna, and that you are not setting `verifyAccess: true` in your lerna.json or `--verify-access` on the CLI otherwise you will be opting into the old behaviour which will not work with automation tokens.
